FHEM/14_SD_UT.pm

```diff
@@ -187,13 +231,13 @@ sub SD_UT_Set($$$@) {
 			IOWrite($hash, 'sendMsg', $msg);
 		}
 	############ Westinghouse Buttons_five ############
-	} elsif 	($model eq "Buttons_five") {
+	} elsif ($model eq "Buttons_five") {
```

Comment:
Ich finde solche IF / Else Schlangen immer sehr unübersichtlich.

Was wäre mit einem Hash in dem für jedes Modell ein Key hinterlegt ist. Dem Key könnte man problemlos eine SUB oder auch mehrere zuweisen.

Ähnlich dem SD_WS Modul

Reply:
Gern schaue ich mir mal die Variante an mit den SUB. Mal schauen ob ich es hinbekomme.

Reply:
Ich schaute es mir nochmal an.
>Was wäre mit einem Hash in dem für jedes Modell ein Key hinterlegt ist. Dem Key könnte man problemlos eine SUB oder auch mehrere zuweisen.

Ähnlich dem SD_WS Modul

Da wir nicht jedem Modell einen einzigen eindeutigen Key zuweisen können würde ich diese Weise beibehalten wollen.

Bsp nicht eindeutige Keys:
- wenn 2 Protokoll ID zutreffen anstatt nur einer eindeutigen

Bei nur eindeutigen Werten verstehe ich deine Denkweise. Da wir in dem Modul nicht um eine „Mischvariante“ drum herum kämen, so würde ich mit den eingefügten Texttrennern das beibehalten.

Dort wo es angebracht ist, orientiere ich mich schon nach Hash / keys Varianten.

Reply:
Mir war nicht bewusst, dass es Varianten gibt, bei denen das Modell nicht eindeutig ist.

Ich habe den Quelltext nicht vor Augen, aber könnte man nicht die Protokolle als Unterttyp des Modells betrachten?

Reply:
Mir war nicht bewusst, dass es Varianten gibt, bei denen das Modell nicht eindeutig ist.

Ich habe den Quelltext nicht vor Augen, aber könnte man nicht die Protokolle als Unterttyp des Modells betrachten?

Reply:
>Mir war nicht bewusst, dass es Varianten gibt, bei denen das Modell nicht eindeutig ist.

Das ist leider der Fall wenn 2 Protokolle zutreffen beim Dispatch innerhalb eines Modules. So muss man dies gegenprüfen sonst erhält man ständig neue Devices.

>... Protokolle als Untertyp

Wie meinst du dies? Kann mir gerade da nichts vorstellen.

Reply:
@sidey79 
nach ausgiebigen hin und her überlegen, so komme ich nicht auf einen Nenner wenn ich einen Teil des Modules in Subs stecken würde oder in einen Hash mit Untertypen. Ich stoße immer wieder auf die Argumente, 
- keine EINDEUTIGE "Klasse" weil 2 ID´s zutreffen innerhalb eines Modules und ich da jeweils 2 DEF Prüfungen vornehmen muss
- viele Prüfungen beziehen sich auf das Attribut Model welches nicht eindeutig erkennbar wäre anhand einer ID weil 2 ID´s bei einer Message zutreffen

kurzum, ich denke wenn ich aufgrund der Situationen das ganez mit "2 IF´s Abfragen" eingrenzen kann innerhalb einer "IF Schlange" so sehe ich den Code besser übersichtlicher.

Reply:
@HomeAutoUser 

Wenn Du sagst, dass es nicht klappt, glaube ich das auch. Du hast dich intensiver beschäftigt als ich.


Ich sehe im code halt if /else Abfragen auf Modell:
```
unknown
Buttons_five
CAME_TOP_432EV
QUIGG_DMV, RH787T
Novy_Pureline_6830, SA_434_1_mini
Unitec_47031
Unitec_other
```

Wenn ich mich nicht vertan habe, dann gibt es immer genau einen else Zweig pro Modell.
Daher nehme ich an, dass man einen Hash erstellen könnte

```
my %modellSubs  = ( 'Button_five' => sub {
    
     return if ($protocol != 29 && $protocol != 30) ;


		$model = AttrVal($name, "model", "unknown");
		$state = substr($bitData,0,6);
		$deviceCode = substr($bitData,8,4);

		## Check fixed bits
		my $unknown1 = substr($bitData,6,1);	# every 1
		my $unknown2 = substr($bitData,7,1);	# every 1
		if ($unknown1 ne "1" | $unknown2 ne "1") {
			Log3 $name, 3, "$ioname: $model fixed bits wrong! always bit6=1 ($unknown1) and bit7=1 ($unknown2)";
			return "";
		}

		## deviceCode conversion for User in ON or OFF ##
		my $deviceCodeUser = $deviceCode;
		$deviceCodeUser =~ s/1/off|/g;
		$deviceCodeUser =~ s/0/on|/g;
		$deviceCodeUser = substr($deviceCodeUser, 0 , length($deviceCodeUser)-1);
		$deviceCode = $deviceCode." ($deviceCodeUser)";

		Log3 $name, 4, "$ioname: $model devicecode=$deviceCode state=$state ($rawData)";
		if ($state eq "011111") {
			$state = "fan low speed";
		} elsif ($state eq "111111") {
			$state = "fan medium speed";
		} elsif ($state eq "111101") {
			$state = "fan high speed";
		} elsif ($state eq "101111") {
			$state = "light on_off";
		} elsif ($state eq "111110") {
			$state = "fan off";
		} else {
			Log3 $name, 3, "$ioname: SD_UT $model unknown Button receive ($state) or houscode is double!";
			return "";
		}
		
		#$state.= " | ".TimeNow();
            }
}
)

Reply:
@sidey79  Es wird eine Codeumstellung geben. Nach intensiven probieren und studieren stellt es auch eine Verienfachung dar. Wir werden ein Hash erzeugen wo wir auf set Befehle zugreifen und auch die Bits dort sofort herbeziehen. 

```
my %models3 = (
# keys => values
  "Buttons_five" => { 	"011111" 	=> "1_fan_low_speed",
					"111111" 	=> "2_fan_medium_speed",
					ID 		=> "P29",
					Type 	=> "remote"
				},
  "CAME_TOP_432EV" => {	"111" => "left_button",
						"010" => "right_button",
						ID => "12458",
						AD => "xse",
						 DA => "dse"
					},
  "unknown" => {	"" => ""
			}
);
```

Dadurch können wir etliche "IF Schlangen" auflösen und es stellt eine sehr deutliche optimierung her, so das wir auch mit neuen Devices uns die Arbeit erleichtern. ;-)

Reply:
In der Parse Funktion in dem Modul geht auch noch was. Wir können das HASH somit in SET und in der Parse "ausreizen" das auch somit in der Parse CODE verschwindet :-)

```
		if ($state eq "110111") {					# button 1
			$state = "1 - fan minimum speed";
		} elsif ($state eq "110101") {				# button 2
			$state = "2 - fan low speed";
		} elsif ($state eq "101111") {				# button 3
			$state = "3 - fan medium low speed";
		} elsif ($state eq "100111") {				# button 4
			$state = "4 - fan medium speed";
		} elsif ($state eq "011101") {				# button 5
			$state = "5 - fan medium high speed";
		} elsif ($state eq "011111") {				# button 6
			$state = "6 - fan high speed";
		} elsif ($state eq "111101") {				# button Turn the fan off
			$state = "Turn the fan off";
		} elsif ($state eq "111110") {				# button light on/off
			$state = "light on/off";
		} elsif ($state eq "111011") {				# button fan direction
			$state = "fan direction";
		} elsif ($state eq "101101") {				# button in remote backside
		$state = "set";
		} else {
			Log3 $name, 3, "$ioname: SD_UT $model unknown Button receive ($state) or houscode is double!";
			return "";
		}
```

so etwas verschwindet dann auch und wird durchlaufen mit einer Schleife.
Du kannst ja dann mal alt gegen NEU vergleichen :-D aber wir stellen es erstmal fertig.